src/executables/csql.c

```diff
@@ -1460,6 +1464,205 @@ csql_do_session_cmd (char *line_read, CSQL_ARGUMENT * csql_arg)
 	}
 
       break;
+
+    case S_CMD_CONNECT:
+      /*dbname_cpy can be stored DB name+ @ + remote hostname + \0 */
+      char dbname_cpy[MAXLOGNAME + CUB_MAXHOSTNAMELEN + 2];
+      char username_cpy[DB_MAX_USER_LENGTH + 1];
+      char host_cpy[DB_MAX_USER_LENGTH + 1];
+      const char *delim = " @\n";
+      char *save_ptr_strtok = NULL;
+      char *db_name = NULL;
+      char *user_name = NULL;
+      char *host_name = NULL;
+      char *p = NULL;
+      CSQL_ARGUMENT csql_new_arg;
+
+      /*find @hostname */
+      host_name = strchr (argument, '@');
+      if (host_name != NULL)
+	{
+	  if (*(host_name + 1) == '\0')
+	    {
+	      host_name = NULL;
+	      csql_Error_code = CSQL_ERR_CONNECT;
+	      return DO_CMD_FAILURE;
+	    }
+	  else
+	    {
+	      host_name += 1;
+	    }
+	}
+
+      if (host_name != NULL)
+	{
+	  if (strcasecmp ("127.0.0.1", host_name) == 0 || strcasecmp ("localhost", host_name) == 0
+	      || strcasecmp (boot_get_host_name (), host_name) == 0 || strcasecmp (boot_get_ip (), host_name) == 0)
+	    {
+	      *(host_name - 1) = '\0';
+	    }
+	}
+
+      strcpy (dbname_cpy, argument);
+      strcpy (username_cpy, argument);
+      /*find db name following the user name */
+      db_name = strrchr (dbname_cpy, ' ');
+      if (db_name != NULL)
+	{
+	  if (*(db_name + 1) == '\0')
+	    {
+	      db_name = NULL;
+	      csql_Error_code = CSQL_ERR_CONNECT;
+	      return DO_CMD_FAILURE;
+	    }
+	  else
+	    {
+	      db_name += 1;
+	    }
+	}
+      user_name = strtok_r (username_cpy, delim, &save_ptr_strtok);
+      if (user_name == NULL)
+	{
+	  csql_Error_code = CSQL_ERR_CONNECT;
+	  return DO_CMD_FAILURE;
+
+	}
+
+      memset (&csql_new_arg, 0, sizeof (CSQL_ARGUMENT));
+      csql_arg_copy (&csql_new_arg, csql_arg);
+
+      /*connect to same DB user or not */
+      if ((db_name == NULL || strcasecmp (csql_arg->db_name, db_name) == 0) && csql_Database_connected == TRUE)
+	{
+
+	  if (!au_is_dba_group_member (Au_user))
+	    {
+
+	      error_code = au_login (user_name, NULL, false);
+	      if (csql_Is_interactive && db_error_code () == ER_AU_INVALID_PASSWORD)
+		{
+		  p = getpass ((char *) csql_get_message (CSQL_PASSWD_PROMPT_TEXT));
+		  error_code = au_login (user_name, p, false);
+		}
+	    }
+	  else
+	    {
+	      error_code = au_login (user_name, NULL, false);
+	    }
+	  if (error_code == NO_ERROR)
+	    {
+	      error_code = clogin_user (user_name);
+	    }
+	  else
+	    {
+	      return DO_CMD_FAILURE;
+	    }
+	  if (error_code == NO_ERROR)
```

Comment:
It is recommended to put a space or two for readability.

Reply:
I will apply it.